Cpx SV PR series, part-6: Update prototyping SV discovery manager tool

SHuang-Broad opened the pull request:
As more and more branches are cleaned up, trunk needs to be updated.

This reflects updates to the tree described in #2703

magicDGS commented on src/test/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/ChimericAlignmentUnitTest.java:
```diff
@@ -7,8 +7,6 @@
 import htsjdk.samtools.SAMRecord;
```
This files might have conflicts after https://github.com/broadinstitute/gatk/pull/3475

SHuang-Broad replied:
Yep. Thanks for pointing out.
I do plan to go over the SV tests when I have spare time over the weekend, and update them as appropriate.
Maybe I'll even name you as the reviewer/victim :smirk:

magicDGS replied:
For the test code I'm happy to review as I did the refactor of the base test for my own benefit; I do not think that I will be useful for the rest of the SV code if you mean that!

SHuang-Broad replied:
definitely not the sv code. just those 1-liner changes that change `BaseTest` to `GATKBaseTest`, if such changes are necessary (i.e. if not already done). 
Thanks!

cwhelan commented:
@SHuang-Broad can you resolve conflicts on this before review?

SHuang-Broad commented:
@cwhelan Here you go.
Just cleaned up and renamed this batch of series of PR that should enable us to call ALL types of SV based on alignment signature alone.
Now we just need to wait for feedback on if the call format makes sense/is intuitive or not, and go from there.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
I'd refrain from calling things MEIs in the code until we know that's what they are. I can see why you might because most of the things that look like this will be MEIs in practice, but there will be other cases where it's not an MEI and it's confusing to call it an MEI before we've verified it (ie checked the insertion sequence against an MEI database or annotation set.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
You put this set into the map here under the `TandemDupOrMEIBkpt` key, but then a few lines later you override it with a different set -- why? I'd recommend not putting collections into the map until they're complete for clarity.

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
I'd prefer a name like `doesNotSpanDuplicationBreakpoints` or something similar.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
This assumption seems weird given the method name and what it does. I'd remove it.

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
Given the method names, maybe the right thing to do here is to just return false if the predicate doesn't hold? Ie return false if the mappings are on different chromosomes.

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
This message seems wrong, and again, it might be better to just return false if the condition doesn't hold rather than throwing an error, given how you've structured the code and named these methods.

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
guaranteed

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
Why do you need to create this dummy object? If it's just to get access to an instance method of `ChimericAlignment` I'd rather refactor that method to be static and just call it from here without needing to create dummy field values, etc.

SHuang-Broad replied:
replaced with the now public method `ChimericAlignment.isLikelySimpleTranslocation(...)`

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
I don't think you need this as a separate static variable, I'd just inline it.

SHuang-Broad replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
Maybe call this `ambiguousMappingsToBeRescued` or something like that?

SHuang-Broad replied:
removed this and the next method but kept TODO's.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
Remove this until there's an implementation.

SHuang-Broad replied:
removed

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/ChimericAlignment.java:
```diff
@@ -303,7 +309,9 @@ static boolean isLikelySimpleTranslocation(final AlignmentInterval regionWithLow
      *  <li>intra-chromosomal translocation that DOES NOT involve a strand switch, i.e.
-     *      novel adjacency between reference locations on the same chromosome involving NO strand switch but reference interval order switch
+     *      novel adjacency between reference locations on the same chromosome involving NO strand switch,
+     *      but in the meantime, the two inducing alignments CANNOT overlap each other since that would point to
```
"but in the meantime," -> "with the condition that"

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
```diff
@@ -103,6 +104,12 @@ private static void decisionTreeForContigsWithTwoAlignments(final JavaRDD<Aligne
 
     // TODO: 10/27/17 add tests for these predicates
 
+    static boolean hasIncompletePicture(final AlignedContig contigWithOnlyOneConfigAnd2Aln) {
```
Again see my other comment about renaming this to be more clear about what's going on, ie something like `doesNotSpanDuplicationBreakpoints` or something

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
```diff
@@ -97,7 +97,7 @@ private static void decisionTreeForContigsWithTwoAlignments(final JavaRDD<Aligne
                 RDDUtils.split(strandSwitchOrNot._2, AssemblyContigAlignmentSignatureClassifier::indicatesIntraChrTandemDupBkpts, false);
         contigsByRawTypes.put(RawTypes.InsDel, tandemDupBkptOrSimpleInsDel._2);
-        contigsByRawTypes.put(RawTypes.TandemDupOrMEIBkpt, tandemDupBkptOrSimpleInsDel._1);
+        contigsByRawTypes.replace(RawTypes.TandemDupOrMEIBkpt, contigsByRawTypes.get(RawTypes.TandemDupOrMEIBkpt).union(tandemDupBkptOrSimpleInsDel._1));
```
Oh, I see, you fixed this later. I think it might be clearer to keep the first set of alignments in a local variable, and then then only add it to the map after joining the two groups together.

SHuang-Broad replied:
done as suggested

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/AssemblyContigAlignmentSignatureClassifier.java:
```diff
@@ -136,6 +136,7 @@ static boolean hasIncompletePictureDueToOverlappingRefOrderSwitch(final AlignedC
         if (referenceSpanOne.contains(referenceSpanTwo) || referenceSpanTwo.contains(referenceSpanOne))
             return true;
 
+        // TODO: 10/29/17 this obsoletes the inverted duplication call code we have now, but those could be used to figure out how to annotate which known ref regions are invert duplicated 
```
This TODO comment is also not very clear to me. Can you explicitly write down what needs to be done?
